### DISCUSSION TITLE
Howto create a Trigger Activity

### DISCUSSION THREAD

**Discussion by robverhoef-united** [elsa-workflows/elsa-core#4891] [Q&A]
Hello,

What is needed to create a trigger activity like the HttpEndpoint or Schedule Activity. As i understand you have to derive it from the ITrigger interface. But other than that... So what are the specifics? Like i want to start a workflow with to be developed FileWatcher activity? How do i build this activity and add it to the system..

Thanks

**Comment by glime-ai** [comment]
To create a custom trigger activity like `HttpEndpoint` or `Schedule` that can start a workflow, you need to implement the `ITrigger` interface in your activity class. Here are the steps you would typically follow to create and integrate a custom `FileWatcher` trigger activity into Elsa Workflows:

1. **Implement the `ITrigger` Interface:**
   Your `FileWatcher` activity class should inherit from `Activity` and implement the `ITrigger` interface. This interface requires you to implement the `GetTriggerPayloadsAsync` method, which provides payloads that can be used to start workflows.

2. **Create the Activity Class:**
   Define the `FileWatcher` class with the necessary properties to configure the file watching behavior (e.g., directory path, filter, etc.). Implement the `Execute` method to create a bookmark that will be used to resume the workflow when a file event occurs.

3. **Handle File Events:**
   Within your `FileWatcher` activity, you will need to set up the actual file watching logic using `FileSystemWatcher` or a similar mechanism. When a file event that you want to trigger the workflow occurs, use the `WorkflowExecutionContext` to resume the workflow using the bookmark created earlier.

4. **Register the Activity:**
   To make your custom activity available to Elsa, you need to register it with the Elsa services. This can be done in the `Startup.cs` or `Program.cs` file of your application, where you configure the Elsa services.

5. **Use the Activity in Workflows:**
   Once registered, you can use your `FileWatcher` activity in workflows. You can do this programmatically by adding it to a workflow definition or by using Elsa Designer if you expose your activity to the designer.

Here's a simplified example of what the `FileWatcher` activity class might look like:

```csharp
public class FileWatcher : Activity, ITrigger
{
    // Define properties for directory path, filter, etc.

    protected override void Execute(ActivityExecutionContext context)
    {
        // Create a bookmark to suspend the workflow
        context.CreateBookmark();

        // Set up file watching logic and handle events
        // ...
    }

    public ValueTask<IEnumerable<object>> GetTriggerPayloadsAsync(TriggerIndexingContext context)
    {
        // Return payloads if necessary for the trigger
        return new ValueTask<IEnumerable<object>>(Enumerable.Empty<object>());
    }
}
```

And to register the activity:

```csharp
services.AddE

**Reply by robverhoef-united** [reply]
Hello, 
Thanks for the answer. I already found a simular example on the elsav3 page, but still having trouble to get this going. I've made a simple test example with a Task.Delay(1000) which should start this workflow every 1 second. My activity looks like this

```
public class TestTrigger : Activity, ITrigger
{
    private ILogger<TestTrigger> _logger;
    private IWorkflowRuntime _workflowRuntime;

    protected override async ValueTask ExecuteAsync(ActivityExecutionContext context)
    {
        _logger = context.GetRequiredService<ILogger<TestTrigger>>();
        _workflowRuntime = context.GetRequiredService<IWorkflowRuntime>();
        _logger.LogInformation("Trigger Test Activity Started");

        context.CreateBookmark();

        trigger(context);
    }

    public ValueTask<IEnumerable<object>> GetTriggerPayloadsAsync(TriggerIndexingContext context)
    {
        //_workflowRuntime = context.ExpressionExecutionContext.ServiceProvider.GetRequiredService<IWorkflowRuntime>();
        //_workflowRuntime.StartWorkflowAsync(context.WorkflowIndexingContext.Workflow.Identity.DefinitionId, new StartWorkflowRuntimeOptions());
        return new ValueTask<IEnumerable<object>>(Enumerable.Empty<object>());
    }

    private void trigger(ActivityExecutionContext context)
    {
        var b = context.Bookmarks.Where(b => b.ActivityInstanceId == context.Id).FirstOrDefault();
        Task.Delay(TimeSpan.FromSeconds(1));
        var options = new ResumeWorkflowRuntimeOptions()
        {
            BookmarkId = b.Id,
        };

        _logger.LogInformation("Resume Trigger Test Activity");

        _workflowRuntime.ResumeWorkflowAsync(context.WorkflowExecutionContext.Id, options);
    }
}
```
I've noticed during testing that the "GetTriggerPayloadsAsync" is called two times during the start of the application so i though i may have to add code to start the particular workflow from the "TriggerIndexingContext" but that does not seems to do the trick.

This is the test workflow:

```
    public class TestTriggerWorkflow: WorkflowBase
    {
        protected override void Build(IWorkflowBuilder builder)
        {
            builder.DefinitionId = "TestTriggerWorkflow";
            builder.Name = "Test Trigger Workflow";
            builder.Version = 1;
            builder.Root = new Sequence
            {

                Activities = {
                    new TestTrigger
                    {
                        CanStartWorkflow = true,
                    },
                    new WriteLine($"Timer is triggered {DateTime.Now}")
                }
            };
        }
    }
```
And added this to program.cs

```
builder.Services.AddElsa(elsa =>
{
\\\\\\\\\\
\\\\\\\\\\\

    elsa.AddActivity<TestTrigger>();

    elsa.AddWorkflow<TestTriggerWorkflow>();
});
```

I also tested with a httpendpoint in a workflow using `elsa.AddHttp()` and that works perfectly. Every time you call the endpoint a new workflow instance in created and executed.

So what am i missing?

Any help is very much appreciated.

Rob

**Reply by GersonDias** [reply]
I'm trying something similar, did you find out how to implement custom triggers properly?

**Reply by robverhoef-united** [reply]
Tobe honest, no i have not. You might want to take a look at the way triggers are changed and expanded in the current V3.5 though. I have not been able to test these yet.
Regards

**Comment by sfmskywalker** [comment]
`ITrigger` marks an activity as a workflow-start trigger, but it is only one part of a file-watching integration.

For a custom `FileWatcher`, derive from `Trigger` (rather than implementing `ITrigger` directly) and override `GetTriggerPayload` / `GetTriggerPayloadsAsync` to return a stable payload for the configured watch, for example a normalized directory plus filter. When a workflow definition is published, Elsa indexes that payload as a `StoredTrigger`.

The process that watches the file system should be a separate hosted service. When it observes a file event, have it send the same stimulus through `IStimulusSender.SendAsync(activityTypeName, stimulus, metadata)`. Elsa then finds matching published trigger workflows and starts them; the same call can also resume matching bookmarks.

Register the activity with the host (for example, `services.AddActivity<FileWatcher>()`). Avoid putting a long-lived `FileSystemWatcher` inside the activity's execution method: published definitions are indexed metadata, while watching is application-host infrastructure.

Relevant current implementation:

- `Elsa.Workflows.Core/Contracts/ITrigger.cs` — trigger payloads represent bookmark data.
- `Elsa.Workflows.Core/Abstractions/Trigger.cs` — base class for overriding payload creation.
- `Elsa.Workflows.Runtime/Services/TriggerIndexer.cs` — persists each indexed payload as a `StoredTrigger`.
- `Elsa.Workflows.Runtime/Services/StimulusSender.cs` — matches a stimulus to both new workflows and existing bookmarks.

This is based on the current codebase, so the exact namespaces may differ from the 3.2-era API in the original question.
